stardog/connection.py

```diff
@@ -125,32 +125,44 @@ def commit(self):
         self.client.post("/transaction/commit/{}".format(self.transaction))
         self.transaction = None
 
-    def add(self, content, graph_uri=None):
+    def add(self, content, graph_uri=None, server_side=False):
         """Adds data to the database.
 
-        Args:
-          content (Content): Data to add
-          graph_uri (str, optional): Named graph into which to add the data
-
-        Raises:
-          stardog.exceptions.TransactionException
-            If not currently in a transaction
+        :param content: Data to add to a graph.
+        :type content: Content, str
+        :param graph_uri: Named graph into which to add the data.
+        :type graph_uri: str, optional
+        :param server_side: Whether the file to load lives in the remote server.
+        :type server_side: bool
+        :raises stardog.exceptions.TransactionException If not currently in a transaction
 
         Examples:
-          >>> conn.add(File('example.ttl'), graph_uri='urn:graph')
+            Loads example.ttl from the current directory
+            >>> conn.add(File('example.ttl'), graph_uri='urn:graph')
+
+            Loads /tmp/example.ttl existing in the remote stardog server, and loads it in the default graph.
+            >>> conn.add(File('/tmp/example.ttl'), server_side=True)
         """
         self._assert_in_transaction()
 
-        with content.data() as data:
-            self.client.post(
-                "/{}/add".format(self.transaction),
-                params={"graph-uri": graph_uri},
-                headers={
+        args = {"params": {"graph-uri": graph_uri}}
+
+        if server_side:
+            args["headers"] = {
+                "Content-Type": "application/json",
+            }
+            args["json"] = {
+                "filename": content.fname,
+            }
+            self.client.post("/{}/add".format(self.transaction), **args)
```

Comment:
This `post` call is the same in both the `if` and `else` cases, so it can be moved to a single call after the `if/else` block.

Reply:
it's not that simple. moving the post call after the else statement will fail when processing any local file that's not open with the with statement (the post call needs to be inside the with statement, otherwise we will try to access a file that's closed)
I also can't encapsulate the entire `if/else` statement within the with statement (i.e move the with statement to the top) because if server side is passed, then `with content.data() as data` will try to open a remote file, which ends up on a non-existent file error.
I am trying to figure out a better way to use a single call for both cases but I am not having much luck.